### ISSUE TITLE
[scroll-animations-1][css-animations-2] Out-of-range range offsets

### ISSUE THREAD

**Issue by fantasai** [w3c/csswg-drafts#8578]
Currently `animation-range-start/end` take `<length-percentage>` values to indicate a position on the timeline. By allowing `<length>` values in #7575, we made it possible to choose positions outside the declared range; and unless we restrict the percentages to [0%,100%], that can be done with percentages also.

The question here is, what do we do about such out-of-range values?
- Make negative values and percentages outside [0%,100%] invalid, and clamp any other values to the named range.
- Allow all values as syntactically valid, but clamp them to the named range.
- Make negative values and percentages outside [0%,100%] invalid when a range name is not specified, and clamp all values to the timeline.
- Allow all values as syntactically valid, but clamp them to the timeline.
- Allow all values, and allow `animation-range` to expand the attachment range for animations.

I think the last one makes the most sense. See also https://github.com/w3c/csswg-drafts/issues/8405#issuecomment-1464810166

**Comment by fantasai** [comment]
@bramus That applies to keyframes, this is about `animation-range`.

**Comment by flackr** [comment]
I agree we should allow all values and not clamp them. This will allow authors to define animations that start just before or end just after particular ranges.

**Comment by bramus** [comment]
So they should not expand the attachment range and simply get clamped?

**Comment by flackr** [comment]
When you say clamped what do you mean? I just said I don't think we should clamp the values :-)

Note that currently if you try to animate beyond the cover range we will clip the animation to the cover range just as we clip keyframes outside of 0-100% to the active range.

**Comment by bramus** [comment]
Ugh. Read/wrote too fast. Clipping is what I intended.

So, by accepting all values and not clamping, would that stretch out the animation-range or would it clip?

**Comment by flackr** [comment]
It will stretch out the animation range. The thing you're specifying in animation-range is a start and end point on the timeline. There's nothing special about the specified range-name especially since the start and end range-name can be different and cross other ranges. (e.g. entry and exit).

**Comment by ydaniv** [comment]
> Does it depend on https://github.com/w3c/csswg-drafts/issues/8672? Can't you specify all possibilities using the longhands?

If `<percentage>` without a range name is not  computed differently, i.e. expanded to a `<name>` + `<percentage>`, then maybe it doesn't depend on #8672, but I was assuming that for view-timelines a `<length-percentage>` will always be expanded that way.

> Using absolute offsets is already supported by spec using the non-named-range <length-percentage> version of animation-range on a scroll timeline. https://www.w3.org/TR/scroll-animations-1/#animation-range-start
>
> There's an [open bug](https://bugs.chromium.org/p/chromium/issues/detail?id=1441013) to get this working in chromium's implementation.

Yes, we support `<length>` for attaching to an absolute offset on scroll-timelines, but I was hoping we could use existing syntax to also limit effects and mapping their progress to the given layout.

> Maybe we should revisit the range of view-timeline and make it so that it can reference ranges over the entire scroll which I think would directly support your use case, e.g.
>
```css
.target {
  /* Animate from initial scroll position to exit. */
  animation-range-start: 0px;
  animation-range-end: exit 100%;
}
```

Yes, that could be useful. But I realize this should be on a separate issue (:

-----

Back to the issue at hand, if I understand it correctly, the question is: "do we allow range offset values that correspond to outside the entire `cover` range", correct?

So regarding @flackr's comment:

> I agree we should allow all values and not clamp them. This will allow authors to define animations that start just before or end just after particular ranges.

This is specifically the case with effects like:

```css
@keyframe parallax {
  from {
    transform: translateY(-500px);
  }
  to {
    transform: translateY(500px);
  }
}

.parallaxed {
  animation: parallax both view();
  animation-range: cover 0% cover 100%;
}
```

Because we ignore the transforms, the animation will actually start when the element is 500px above `cover 0%` and end when the element is 500px below `cover 100%`.
Now, there are 3 ways to tackle this:

1. Extend the effective scrollport with `view-timeline-inset: -500px`.
2. Allow negative values like suggested here, and set: `animation-range: entry -500px exit 500px` or with calc()'s: `calc(entry 0% - 500px) calc(exit 100% + 500px)`, but I assume these are equivalent, right?
3. Use a named `view-timeline` on a different element that's 1000px higher and the `.parallaxed` element is centered inside it.

Same can be said for effects where the animated element is offset from the view-timeline subject and or extends beyond it.

I think option 3 requires an extra element, so having to resort to that is a less desirable solution.

So, if my conclusion is correct, it's either we clamp for simplicity's sake, or we allow offsets that are out-of-range?
I think we can allow out-of-range if that makes more sense to authors, no particular reason why not to allow it also.

**Comment by flackr** [comment]
> Because we ignore the transforms, the animation will actually start when the element is 500px above cover 0%

Technically the element isn't 500px above cover until the animation starts 😉

I think that either options 1 or 2 should work. Note that the range name can't go inside the calc. Option 2 would be `animation-range: entry -500px exit calc(100% + 500px);`. `exit 500px` is equivalent to `exit calc(0% + 500px)` which is 500px after the *start* of the exit range which isn't what you want in your use case.

> So, if my conclusion is correct, it's either we clamp for simplicity's sake,

I don't think we should clamp. It results in animations which behave differently if the animated element is near the top/bottom of the document versus further down.

> or we allow offsets that are out-of-range?
I think we can allow out-of-range if that makes more sense to authors, no particular reason why not to allow it also.

Yes, I think we should allow out-of-range values.

Note I tested this and it seems the chromium implementation already supports this (requires a very recent version e.g. >=115.0.5767.0):
https://jsbin.com/hitegay/edit?html,css,js,output

**Comment by ydaniv** [comment]
@flackr SGTM.

Another use-case for explicitly clamping, the following can work for ensuring something animates from entry, even if in some screen/layout it appears in first fold, and in others it's below:

```css
.target {
  animation-range-start: max(100vh, entry 0%);
  animation-range-end: contain 50%;
}
```

----

Regarding negative values, if we'll use absolute values without a range name to refer to the entire scroll range, then negative values become practically useless. Perhaps we should consider again using these to specify offsets from the end of the scroll range?

**Comment by flackr** [comment]
> Another use-case for explicitly clamping, the following can work for ensuring something animates from entry, even if in some screen/layout it appears in first fold, and in others it's below:
> 
> ```css
> .target {
>   animation-range-start: max(100vh, entry 0%);
>   animation-range-end: contain 50%;
> }
> ```

I suspect it would be hard to support range names within the calc. Maybe it could be done - since the entry 0% effectively turns into a proportion of the timeline range which is comparable with 100vh - but I think this could get complex.

> Regarding negative values, if we'll use absolute values without a range name to refer to the entire scroll range, then negative values become practically useless. Perhaps we should consider again using these to specify offsets from the end of the scroll range?

I think negative values are still useful when associated with a range name, e.g.
```css
.target {
  /* Start 50px before the contain phase */
  animation-range-start: contain -50px;
}
```

**Comment by ydaniv** [comment]
D'oh! I messed up the example, took screen end instead of start, should be:

```css
.target {
  animation-range-start: max(0%, entry 0%);
  animation-range-end: contain 50%;
}
```

> I suspect it would be hard to support range names within the calc. Maybe it could be done - since the entry 0% effectively turns into a proportion of the timeline range which is comparable with 100vh - but I think this could get complex.

I guess this requires transforming range names into functions, and then they could be simply parsed as `<length>`, right?
Is there another option?

> I think negative values are still useful when associated with a range name, e.g.
> ```css
> .target {
>   /* Start 50px before the contain phase */
>   animation-range-start: contain -50px;
> }
> ```

Yes of course, it's useful for names in that way, but useless for absolute values.

> While we could do something different when no range name is given, I worry that it would be very magical, and also make it harder to work with calcs. E.g. imagine your calc sometimes produces a negative value - it shouldn't suddenly flip to the end of the range.

Yep, that's a real concern. I guess this is something that can be added later if demand arises.

**Comment by bramus** [comment]
Looking at the original issue, this was about allowing values greater than 100% and less than 0%. Maybe we should split off the min/max functionality to a separate issue?
